src/closure-types.ts

```diff
@@ -96,20 +104,42 @@ function nullableByDefault(node: doctrine.Type): boolean {
   return isArray(node);
 }
 
+/**
+ * Return whether the given node should be wrapped in parenthesis when it is an
+ * argument to a greedy operator. For example, if we are serializing a union,
+ * and one of the arguments is itself a union, then that argument should be
+ * wrapped in parenthesis to avoid precendence problems.
+ */
+function ambiguousPrecendence(node: doctrine.Type): boolean {
+  if (isFunction(node)) {
+    return true;
+  }
+  if (isUnion(node)) {
+    if (node.elements.length === 1) {
+      // Unions of length 1 get collapsed, so recurse in case we hit a
+      // descendent that is ambiguous.
+      return ambiguousPrecendence(node.elements[0]);
+    }
+    return true;
+  }
+  return false;
+}
+
 function serializeArray(node: doctrine.type.TypeApplication): string {
   if (node.applications.length !== 1) {
     console.error('Invalid array expression.');
     return '';
   }
-  return serialize(node.applications[0]) + '[]';
+  const arg = node.applications[0];
+  return serialize(arg, true) + '[]';
```

Comment:
Consider just doing `Array<Foo>` in all cases. It's equivalent and scales better to complicated types.

Reply:
I like it, simplifies parenthesis handling too. Done.